crates/bevy_ecs/src/observer/builder.rs

```diff
@@ -0,0 +1,238 @@
+    pub fn on_event_ids(
+        &mut self,
+        events: impl IntoIterator<Item = ComponentId>,
+    ) -> &mut ObserverBuilder<'w, ()> {
+        self.descriptor.events.extend(events);
+        // SAFETY: () type will not allow bad memory access as it has no size
+        unsafe { std::mem::transmute(self) }
+    }
+
+    /// Add [`ComponentId`] in `T` to the list of components listened to by this observer.
+    /// For examples an `OnRemove` observer would trigger when any component in `B` was removed.
+    pub fn components<B: Bundle>(&mut self) -> &mut Self {
+        B::get_component_ids(self.commands.components(), &mut |id| {
+            self.descriptor.components.push(id.unwrap_or_else(|| {
+                panic!(
+                    "Cannot observe event before it is registered with init_component: {}",
+                    std::any::type_name::<B>(),
+                )
+            }));
+        });
+        self
+    }
+
+    /// Add `ids` to the list of component sources listened to by this observer.
+    pub fn component_ids<'c>(
+        &mut self,
+        ids: impl IntoIterator<Item = &'c ComponentId>,
+    ) -> &mut Self {
+        self.descriptor.components.extend(ids.into_iter().cloned());
+        self
+    }
+
+    /// Adds `source` as the list of entity sources listened to by this observer.
+    pub fn source(&mut self, source: Entity) -> &mut Self {
+        self.descriptor.sources.push(source);
+        self
+    }
+
+    /// Spawns the resulting observer into the world.
+    pub fn run<B: Bundle, M>(&mut self, system: impl IntoObserverSystem<E, B, M>) -> Entity {
+        B::get_component_ids(self.commands.components(), &mut |id| {
+            self.descriptor.components.push(id.unwrap_or_else(|| {
+                panic!(
+                    "Cannot observe event before it is registered with init_component: {}",
+                    std::any::type_name::<B>(),
+                )
+            }));
+        });
+        let entity = self.commands.spawn_empty().id();
+        let descriptor = self.descriptor.clone();
+        self.commands.add(move |world: &mut World| {
+            let component = ObserverComponent::from(world, descriptor, system);
+            world.entity_mut(entity).insert(component);
+            world.register_observer(entity);
+        });
+        entity
+    }
+
+    /// Spawns the resulting observer into the world using an [`ObserverRunner`] callback.
+    /// This is not advised unless you want to override the default runner behaviour.
+    pub fn runner(&mut self, runner: ObserverRunner) -> Entity {
+        let entity = self.commands.spawn_empty().id();
+        let descriptor = self.descriptor.clone();
+        self.commands.add(move |world: &mut World| {
+            let component = ObserverComponent::from_runner(descriptor, runner);
+            world.entity_mut(entity).insert(component);
+        });
+        entity
+    }
```

Comment:
Missing `world.register_observer` in comparison to `run`, worth extracting a function for common code